### PR TITLE
[GH-16199] add IsValid to model OpenDialogRequest

### PR DESCRIPTION
**Pull request by annaos**
#### Summary
Add isValid method for openDialog request.

I would like to draw attention here to the error in the [documentation](https://developers.mattermost.com/integrate/plugins/interactive-dialogs/). In the table in the 'Select elements' section, the rows 'options' and 'help_text' are incorrect.
[PR to fix it](https://github.com/mattermost/mattermost-developer-documentation/pull/1348)

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/16199
Jira https://mattermost.atlassian.net/browse/MM-30075

#### Release Note

```release-note
Add validation that the payload for a open Interactive Dialog request is valid according to the rules at https://docs.mattermost.com/developer/interactive-dialogs.html
```

### CODE REVIEW

File: server/channels/app/integration_action.go [hunk 1]
```diff
@@ -460,6 +460,10 @@ func (a *App) OpenInteractiveDialog(request model.OpenDialogRequest) *model.AppE
 		return appErr
 	}
 
+	if dialogErr := request.IsValid(); dialogErr != nil {
+		return model.NewAppError("OpenInteractiveDialog", "app.open_interactive_dialog.invalid_request", nil, dialogErr.Error(), http.StatusBadRequest)
+	}
```
Comment (hanzei) [inline server/channels/app/integration_action.go:465]: Let's wrap the error
```suggestion
	if err := request.IsValid(); err != nil {
		return model.NewAppError("OpenInteractiveDialog", "app.open_interactive_dialog.invalid_request", nil, "", http.StatusBadRequest).Wrap(err)
	}
```

File: server/public/model/integration_action.go [hunk 2]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+	if d.Title == "" || len(d.Title) > DialogTitleMaxLength {
+		multiErr = multierror.Append(multiErr, errors.New("invalid dialog title"))
+	}
+
+	if d.IconURL != "" && !IsValidHTTPURL(d.IconURL) {
+		multiErr = multierror.Append(multiErr, errors.New("invalid icon url"))
+	}
+	if len(d.Elements) > DialogMaxElements {
+		multiErr = multierror.Append(multiErr, errors.Errorf("maximum number of dialog elements is %d", DialogMaxElements))
+	}
```
Comment (hanzei) [inline server/public/model/integration_action.go:379]: That is actually no longer true https://github.com/mattermost/mattermost-developer-documentation/pull/1341

File: server/public/model/integration_action.go [hunk 3]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
 	return DecodeAndVerifyTriggerId(r.TriggerId, s, timeout)
 }
 
+func (r *OpenDialogRequest) IsValid() error {
+	var multiErr *multierror.Error
+	if r.URL == "" || !IsValidHTTPURL(r.URL) {
+		multiErr = multierror.Append(multiErr, errors.New("invalid URL"))
+	}
```
Comment (hanzei) [inline server/public/model/integration_action.go:354]: Plugins can use an absolute path instead of a URL, e.g. `/plugins/com.github.matterpoll.matterpoll/api/v1/polls/somepollid/delete/confirm` (https://github.com/matterpoll/matterpoll/blob/fc3e3e8fc5ec79a41bd7105533d53ec0450ac7d4/server/plugin/api.go#L710). The code should also allow that.

File: server/public/model/integration_action.go [hunk 4]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+
+	if r.TriggerId == "" {
+		multiErr = multierror.Append(multiErr, errors.New("empty trigger id"))
+	}
+
+	err := r.Dialog.IsValid()
+	if err != nil {
+		multiErr = multierror.Append(multiErr, err)
+	}
+	return multiErr.ErrorOrNil()
```
Comment (hanzei) [inline server/public/model/integration_action.go:364]:
```suggestion
	}
	
	return multiErr.ErrorOrNil()
```

File: server/public/model/integration_action.go [hunk 5]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+	}
+	if len(d.Elements) > DialogMaxElements {
+		multiErr = multierror.Append(multiErr, errors.Errorf("maximum number of dialog elements is %d", DialogMaxElements))
+	}
+
+	if len(d.Elements) != 0 {
+		elementMap := make(map[string]bool)
+
+		for _, element := range d.Elements {
+			if elementMap[element.Name] {
```
Comment (hanzei) [inline server/public/model/integration_action.go:385]: Neat implementation :+1:

File: server/public/model/integration_action.go [hunk 6]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+				multiErr = multierror.Append(multiErr, errors.Wrapf(err, "%q field is not valid", element.Name))
+			}
+		}
+	}
+	return multiErr.ErrorOrNil()
+}
+
+func (e *DialogElement) IsValid() error {
+	var multiErr *multierror.Error
+	subTypeMap := map[string]bool{
```
Comment (hanzei) [inline server/public/model/integration_action.go:401]: Maybe `textSubTypes` is more clear.

File: server/public/model/integration_action.go [hunk 7]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+		"text":     true,
+		"email":    true,
+		"number":   true,
+		"tel":      true,
+		"url":      true,
+		"password": true,
+	}
+
+	if e.MinLength < 0 {
+		multiErr = multierror.Append(multiErr, errors.New("min length cannot be a negative number"))
```
Comment (hanzei) [inline server/public/model/integration_action.go:412]: In a more general term, it would be nice if the error messages would contain the falty value so it's easier to spot what is wrong with the values, e.g.
```suggestion
		multiErr = multierror.Append(multiErr, errors.Errorf("min length cannot be a negative number: %q", e.MinLength))
```

File: server/public/model/integration_action_test.go
Comment (hanzei) [inline server/public/model/integration_action_test.go:1]: Great work on the tests :+1: 

I wonder if it makes sense to use a table-driven test as that makes it quite clear which input should produce which output. (https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) Currently, it is hard to spot in which state `request` is (which is why you had to use `t.Cleanup` a lot)

Let me know what you think.

File: server/public/model/integration_action.go [hunk 8]
```diff
@@ -335,6 +347,130 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+
+	default:
+		multiErr = multierror.Append(multiErr, errors.Errorf("invalid element type: %q", e.Type))
+	}
+
+	return multiErr.ErrorOrNil()
+}
+
+func checkMaxLength(fieldName string, field string, length int) error {
+	var check bool
```
Comment (hanzei) [inline server/public/model/integration_action.go:461]: Nit: maybe `valid`?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks @annaos, awesome work with translating the docs into code.

File: server/channels/app/integration_action.go (hunk 1 above)
Comment (annaos) [reply server/channels/app/integration_action.go]: Done. I wonder, if the wrapped information will be used, because for example in `server/channels/api4/integration_action_test.go` in the case `Should fail with same elements` it comes without wrapped error.

File: server/public/model/integration_action_test.go
Comment (annaos) [reply server/public/model/integration_action_test.go:1]: I agree, using `t.Cleanup` was not the best idea. I don't think a table-driven test would be the best solution here because I need similar `OpenDialogRequest` instances, and writing out the whole object for each test case would involve a lot of duplication.

I modified the tests by using a separate instance of `OpenDialogRequest` in each case, so I don't need  `t.Cleanup` anymore

File: server/public/model/integration_action_test.go [hunk 9]
```diff
@@ -180,3 +181,137 @@ func TestPostActionIntegrationEquals(t *testing.T) {
...
+					Type:        "text",
+					Placeholder: "Enter a value",
+				},
+			},
+			SubmitLabel:    "Submit",
+			NotifyOnCancel: false,
+			State:          "somestate",
+		},
+	}
+}
```
Comment (hanzei) [inline server/public/model/integration_action_test.go:317]: Nit pick: this function can be part of `TestOpenDialogRequestIsValid`

File: server/public/model/integration_action.go [hunk 10]
```diff
@@ -361,21 +360,15 @@ func (r *OpenDialogRequest) IsValid() error {
...
 func (d *Dialog) IsValid() error {
 	var multiErr *multierror.Error
 
 	if d.Title == "" || len(d.Title) > DialogTitleMaxLength {
-		multiErr = multierror.Append(multiErr, errors.New("invalid dialog title"))
-	}
-
-	if d.IconURL != "" && !IsValidHTTPURL(d.IconURL) {
-		multiErr = multierror.Append(multiErr, errors.New("invalid icon url"))
-	}
```
Comment (hanzei) [inline server/public/model/integration_action.go]: Why was this removed? `IconURL` must be a URL and can't be a plugin path.

File: server/public/model/integration_action_test.go
Comment (hanzei) [reply server/public/model/integration_action_test.go:1]: Make sense :+1:

File: server/public/model/integration_action.go (hunk 10 above)
Comment (annaos) [reply server/public/model/integration_action.go]: Done. Misunderstanding

**Review by hanzei — APPROVED** [review]
Awesome 🚀

**Comment by mickmister** [comment]
@hanzei Do you think this needs to go through manual QA testing?

File: server/public/model/integration_action.go [hunk 11]
```diff
@@ -335,6 +346,128 @@ func (r *OpenDialogRequest) DecodeAndVerifyTriggerId(s *ecdsa.PrivateKey, timeou
...
+	case "textarea":
+		multiErr = multierror.Append(multiErr, checkMaxLength("Default", e.Default, DialogElementTextareaMaxLength))
+		multiErr = multierror.Append(multiErr, checkMaxLength("Placeholder", e.Placeholder, DialogElementTextareaMaxLength))
+
+		if _, ok := textSubTypes[e.SubType]; !ok {
+			multiErr = multierror.Append(multiErr, errors.Errorf("invalid subtype %q", e.Type))
+		}
+
+	case "select":
+		multiErr = multierror.Append(multiErr, checkMaxLength("Default", e.Default, DialogElementSelectMaxLength))
```
[the remark this answers is not in the document]
Comment (annaos) [reply server/public/model/integration_action.go:436]: Select options can be defined through `data_source`, not special in `options` field

[the remark this answers is not in the document]
Comment (hanzei) [reply server/public/model/integration_action.go:436]: I don't understand your comment @annaos. Would it be reasonable to add a check that ensure the default value exists if  `data_source` is not used?

**Comment by hanzei** [comment]
Yes, let me test the changes against Matterpoll once all discussions are resolved.

File: server/channels/app/integration_action.go [hunk 12]
```diff
@@ -460,6 +460,10 @@ func (a *App) OpenInteractiveDialog(request model.OpenDialogRequest) *model.AppE
 		return appErr
 	}
 
+	if dialogErr := request.IsValid(); dialogErr != nil {
+		return model.NewAppError("OpenInteractiveDialog", "app.open_interactive_dialog.invalid_request", nil, "", http.StatusBadRequest).Wrap(dialogErr)
+	}
```
Comment (hanzei) [inline server/channels/app/integration_action.go:465]: After giving this another thought, I wonder if the code should be more liberal and not error out. In case of bugs in `IsValid` this currently blocks integration developers from open a dialog that the server deems faulty. 

I'm leaning towards logging a warning message instead. @annaos @mickmister what do you think?

Comment (annaos) [reply server/channels/app/integration_action.go:465]: This is a good idea. Errors such as exceeding the maximum length or having the wrong default value are not as critical and should not block the opening of a dialog. However, errors like the wrong element type or an empty URL are critical and will result in a broken dialog, correct?

Should I still return an error for important cases (tbd) and log a warning for the rest? @hanzei @mickmister

Comment (hanzei) [reply server/channels/app/integration_action.go:465]: Good question. I wonder if blocking the dialogue creation at all is helpful. It is not a great UX (as the user doesn't understand why the dialogue won't open and can't fix it).

So I'm leaning towards never blocking the request.

Comment (annaos) [reply server/channels/app/integration_action.go]: I logged an error level message. I hope that's okay with you. I couldn't find a way to check a log message in the test. If you know of one, please let me know. Otherwise everything done.

File: server/channels/api4/integration_action_test.go [hunk 13]
```diff
@@ -209,6 +211,53 @@ func TestOpenDialog(t *testing.T) {
...
+
+		buffer := &mlog.Buffer{}
+		err := mlog.AddWriterTarget(th.TestLogger, buffer, true, mlog.StdAll...)
+		require.NoError(t, err)
+
+		_, err = client.OpenInteractiveDialog(context.Background(), request)
+		require.NoError(t, err)
+
+		require.NoError(t, th.TestLogger.Flush())
+		testlib.AssertLog(t, buffer, mlog.LvlWarn.Name, "Interactive dialog is invalid")
```
Comment (hanzei) [inline server/channels/api4/integration_action_test.go:232]: :100:

**Review by hanzei — APPROVED** [review]
:rocket:

**Comment by hanzei** [comment]
@DHaussermann I suggest you do some smoke testing with the demo plugin and check for any unexpected error messages about invalid dialogs in the logs.

File: server/channels/api4/integration_action_test.go [hunk 14]
```diff
@@ -6,6 +6,8 @@ package api4
 import (
 	"context"
 	"encoding/json"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/v8/channels/testlib"
```
Comment (hanzei) [inline server/channels/api4/integration_action_test.go:10]: golangci doesn't like the order of imports

```
channels/api4/integration_action_test.go:9: File is not `goimports`-ed (goimports)
	"github.com/mattermost/mattermost/server/public/shared/mlog"
	"github.com/mattermost/mattermost/server/v8/channels/testlib"
make: *** [Makefile:290: golangci-lint] Error 1
```

These two imports should be next to the `"github.com/mattermost/mattermost/server/public/model"` import.

If you use VSCode with the Go extension, it automatically formats the import rules.

Comment (annaos) [reply server/channels/api4/integration_action_test.go:10]: I'm using GoLand :cry:

**Review by hanzei — APPROVED** [review]
